### PR TITLE
Multiple sort

### PR DESCRIPTION
**Pull request by jgryko5**

### CODE REVIEW

File: datasette/static/sort_utils.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+          `input[name="${input.name}_t"]:checked`
+        ).value;
+      } catch (e) {
+        alert(`You need to select an option for "${input.name}"!`);
+        return;
+      } finally {
+        let s_option = { name: input.name, value: `${type}` };
+        selected_options.push(s_option);
+      }
+    }
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: nit: reverse this condition and add `continue;` to avoid one level of indentation:
```suggestion
    if (input.type != "checkbox" || input.checked != true) {
      continue;
    }
    try {
      var type = document.querySelector(
        `input[name="${input.name}_t"]:checked`
      ).value;
    } catch (e) {
      alert(`You need to select an option for "${input.name}"!`);
      return;
    } finally {
      let s_option = { name: input.name, value: `${type}` };
      selected_options.push(s_option);
    }
```
Also, avoid using `alert()` as this is blocking the whole browser. You might want to append an error message somewhere in the DOM.

File: datasette/static/sort_utils.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        option["name"]
+      );
+    }
+    //table.js line 37
+    window.location.href = sort_url ? "?" + sort_url : location.pathname;
+  } else {
+    alert("You need to select some columns to sort!");
+  }
+}
+function toggleSortMenu() {
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js:28]: nit: add an empty line before every function to visually separate them and improve the readability of the whole file.

File: datasette/static/sort_utils.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      } catch (e) {
+        alert(`You need to select an option for "${input.name}"!`);
+        return;
+      } finally {
+        let s_option = { name: input.name, value: `${type}` };
+        selected_options.push(s_option);
+      }
+    }
+  }
+  if (selected_options.length > 0) {
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: Same here, if there's a small `else` block, which is the last thing done in this function, reverse the condition and to a `return`. Then you don't need `else` at all. This reads better, as you do all the validation first and interrupt the execution of this function if there are any issues.

File: datasette/static/sort_utils.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    var sort_url = new URLSearchParams(window.search);
+    for (let option of selected_options) {
+      sort_url.append(
+        `_sort${option["value"] == "dsc" ? "_desc" : ""}`,
+        option["name"]
+      );
+    }
+    //table.js line 37
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: Can this comment be a bit more informative?

File: datasette/static/sort_utils.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    });
+  } else {
+    menu.style.display = "none";
+  }
+}
+function populateSortMenu() {
+  var sort_url = new URLSearchParams(window.location.search);
+  let params = Object.fromEntries(sort_url.entries());
+  if (Object.keys(params).length > 0) {
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: Same here, reverse the condition and return.

File: datasette/static/sort_utils.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    for (let [name, param] of Object.entries(params)) {
+      let chb = document.getElementsByName(param)[0];
+      chb.checked = true;
+      document.getElementById(
+        name.includes("_desc") ? `${param}_dsc` : `${param}_asc`
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: nit: assign the id to a variable to make this a bit more readable, using ternary operator here makes it a bit hard to read.

File: datasette/templates/__sorting_example.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<script src="{{ urls.static('sort_utils.js') }}"></script>
+<div>
+    <p style="display:inline-block;">Sorting options:</p>
+    <button type="button" class="btn-secondary" id="toggle_sort_menu" onclick="toggleSortMenu();" style="display:inline-block;">show</button>
+</div>
+<div style="display:block;position: absolute;z-index: 1;">
```
Comment (nineinchnick) [inline datasette/templates/__sorting_example.html]: nit: add spaces after semicolons and colons
```suggestion
<div style="display: block; position: absolute; z-index: 1;">
```

File: datasette/templates/__sorting_example.html [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        <div class="hook"></div>
+        <div class="row" name="sorting_options">
+        <div class="col">
+            <p style="margin:0;">Column | Type</p>
+        </div>
+        {% for column in display_columns if column.sortable %}
+            <div class="col">
+                <input type="checkbox" name="{{column.name}}">
+                {{column.name}} 
+                <input type="radio" id="{{column.name}}_asc" name="{{column.name}}_t" value="asc" checked>
```
Comment (nineinchnick) [inline datasette/templates/__sorting_example.html]: Why there's a `_t` suffix in the name?

File: datasette/utils/__init__.py [hunk 9]
```diff
@@ -70,6 +70,30 @@ async def await_me_maybe(value):
     return value
 
 
+def convert_list_to_sql(list, descending=False):
+    """Takes a Python list, escapes its values and returns it as comma separated values"""
+    value_list = ""
+    for item in list:
+        value_list += f"{escape_sqlite(item)}{' desc' if descending else ''}, "
+    return value_list[:-2]
```
Comment (nineinchnick) [inline datasette/utils/__init__.py]: This should be a list that you'd `join()` with ", " as a separator when returning to get a string
```suggestion
    values = [escape_sqlite(item) + ' desc' if descending else ''
              for item in list]
    return values.join(", ")
```

File: datasette/views/table.py [hunk 10]
```diff
@@ -543,29 +545,38 @@ async def data(
...
-        sort = special_args.get("_sort")
-        sort_desc = special_args.get("_sort_desc")
-
+        #
+        # getting sorting conditions from url and saving them in an list
+        #
+        sort = []
+        sort_desc = []
+
+        if "_sort" in args:
```
Comment (nineinchnick) [inline datasette/views/table.py]: Can this part be simplified? That is, does `args.getlist()` can take a default value to return it when the `_sort` is not set? Also, could you avoid using a loop and just assign the list to `sort`?

File: datasette/views/table.py [hunk 11]
```diff
@@ -543,29 +545,38 @@ async def data(
...
-            sort_desc = table_metadata.get("sort_desc")
-
-        if sort and sort_desc:
-            raise DatasetteError("Cannot use _sort and _sort_desc at the same time")
-
+            if table_metadata.get("sort"):
+                sort.append(table_metadata.get("sort"))
+            if table_metadata.get("sort_desc"):
+                sort_desc.append(table_metadata.get("sort_desc"))
         if sort:
```
Comment (nineinchnick) [inline datasette/views/table.py]: Is this condition required? Would `convert_list_to_sql` produce any output when `sort` is an empty list`?

File: datasette/views/table.py [hunk 12]
```diff
@@ -543,29 +545,38 @@ async def data(
...
         if sort_desc:
-            if sort_desc not in sortable_columns:
-                raise DatasetteError(f"Cannot sort table by {sort_desc}")
-
-            order_by = f"{escape_sqlite(sort_desc)} desc"
+            for column in sort_desc:
+                if column not in sortable_columns:
+                    raise DatasetteError(f"Cannot sort table by {column}")
+            if sort:
+                order_by += ", "
```
Comment (nineinchnick) [inline datasette/views/table.py]: `order_by` should be a list that you later convert to a string using `order_by.join(", ")`.

File: datasette/views/table.py [hunk 13]
```diff
@@ -615,28 +626,36 @@ async def data(
...
+                                    column_null=convert_list_to_sql_is_null(sort_desc),
                                     next_clauses=" and ".join(next_by_pk_clauses),
                                 )
                             )
+
                         else:
                             where_clauses.append(
-                                "({column} is not null or ({column} is null and {next_clauses}))".format(
-                                    column=escape_sqlite(sort),
+                                "({column_not_null} or ({column_null} and {next_clauses}))".format(
```
Comment (nineinchnick) [inline datasette/views/table.py:637]: since `{column_not_null}` can contain multiple expressions, it should be in parenthesis, or `convert_list_to_sql_is_not_null` should always wrap its results in parenthesis.

File: datasette/utils/__init__.py [hunk 14]
```diff
@@ -70,6 +70,30 @@ async def await_me_maybe(value):
     return value
 
 
+def convert_list_to_sql(list, descending=False):
```
Comment (nineinchnick) [inline datasette/utils/__init__.py]: The name is too generic, consider renaming this to `get_order_by`

File: datasette/utils/__init__.py [hunk 15]
```diff
@@ -70,6 +70,30 @@ async def await_me_maybe(value):
...
+    """Takes a Python list, escapes its values and returns it as comma separated values"""
+    value_list = ""
+    for item in list:
+        value_list += f"{escape_sqlite(item)}{' desc' if descending else ''}, "
+    return value_list[:-2]
+
+
+def convert_list_to_sql_is_null(list):
```
Comment (nineinchnick) [inline datasette/utils/__init__.py]: Consider simplifying the name a bit, how about `check_nulls` and `check_not_nulls` for the other one.

File: datasette/views/table.py [hunk 16]
```diff
@@ -615,28 +626,36 @@ async def data(
...
+                                    column_not_null=convert_list_to_sql_is_not_null(
+                                        sort
+                                    ),
+                                    column_null=convert_list_to_sql_is_null(sort),
                                     next_clauses=" and ".join(next_by_pk_clauses),
                                 )
                             )
                     else:
+                        sort_columns = sort if sort else sort_desc
```
Comment (nineinchnick) [inline datasette/views/table.py]: This part can be simplified, using list comprehension, similar to my previous comments.

File: datasette/views/table.py [hunk 17]
```diff
@@ -808,28 +826,31 @@ async def data(
...
+        added_args = []
         if 0 < page_size < len(rows):
             if is_view:
                 next_value = int(_next or 0) + page_size
             else:
                 next_value = path_from_row_pks(rows[-2], pks, use_rowid)
             # If there's a sort or sort_desc, add that value as a prefix
             if (sort or sort_desc) and not is_view:
-                prefix = rows[-2][sort or sort_desc]
+                prefix = rows[-2][sort[0] if sort else sort_desc[0]]
```
Comment (nineinchnick) [inline datasette/views/table.py]: There needs to be a comment here explaining why we take only the first item.

File: datasette/views/table.py [hunk 18]
```diff
@@ -851,11 +872,20 @@ async def data(
...
             extra=extra_human_descriptions
         )
-
-        if sort or sort_desc:
-            sorted_by = "sorted by {}{}".format(
-                (sort or sort_desc), " descending" if sort_desc else ""
-            )
+        sorting_conditions = ""
+        if sort:
+            for i in range(0, len(sort)):
```
Comment (nineinchnick) [inline datasette/views/table.py]: You could use `for i, item in enumerate(sort):`

File: datasette/views/table.py [hunk 19]
```diff
@@ -851,11 +872,20 @@ async def data(
...
+                sorting_conditions += f"{sort[i]}"
+                if i != len(sort) - 1:
```
Comment (nineinchnick) [inline datasette/views/table.py]: Simplify this by building a list and joining it with a separator at the end.

**Review by nineinchnick — CHANGES_REQUESTED** [review]
You're missing new test cases for sorting using multiple columns. Also, I'm not sure but is sorting by multiple columns in different directions supported?

File: datasette/templates/__sorting_example.html (hunk 8 above)
Comment (jgryko5) [reply datasette/templates/__sorting_example.html]: I added the suffix in the radiobuttons names to differentiate them from their checkboxes ("_t" = type of sorting - ascending/descending).

File: datasette/static/app.css [hunk 20]
```diff
@@ -519,7 +519,7 @@ input[type="search"]::-webkit-search-results-decoration {
     display: none;
 }
 
-form input[type=submit], form button[type=button] {
+form input[type=submit], form button[type=button], .btn-primary, .btn-secondary{
```
Comment (nineinchnick) [inline datasette/static/app.css]: nit, leave a blank space before the brace
```suggestion
form input[type=submit], form button[type=button], .btn-primary, .btn-secondary {
```

File: datasette/static/sort_utils.js [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+      while (target && target != menu && target != btn) {
+        target = target.parentNode;
+      }
+      if (!target) {
+        menu.style.display = "none";
+      }
+    });
+  } else {
+    menu.style.display = "none";
+  }
```
Comment (nineinchnick) [inline datasette/static/sort_utils.js]: nit: do an early return to avoid one level of indentation, making this more readable.
```suggestion
  if (menu.style.display != "none") { 
    menu.style.display = "none";
    return;
  }
  menu.style.display = "inline-block";
  menu.classList.add("anim-scale-in");
  //Taken from table.js - lines 79-85
  document.addEventListener("click", function (ev) {
    var target = ev.target;
    while (target && target != menu && target != btn) {
      target = target.parentNode;
    }
    if (!target) {
      menu.style.display = "none";
    }
  });
```

File: datasette/templates/__sorting_example.html [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+                <label for="{{column.name}}_dsc">Descending</label>
+                </div>
+                {% endfor %}
+                <input type="button" class="btn-primary" value="Sort" onclick="sort();" style="width:100%; border-top-right-radius: 0px; border-top-left-radius: 0px;">
+        </div>
+    </div>
+    <script type="text/javascript">
+        populateSortMenu();
+    </script>
+</div>
```
Comment (nineinchnick) [inline datasette/templates/__sorting_example.html]: nit: always add an end-of-line at end-of-file - configure your IDE to do so

File: datasette/utils/__init__.py [hunk 23]
```diff
@@ -116,6 +134,27 @@ def compound_keys_after_sql(pks, start_index=0):
...
+    #
+    # A modified version of compound_keys_after_sql supporting sorting by multiple columns in any direction
+    #
+    or_clauses = []
+    pks_left = sort[:]
+    while pks_left:
+        and_clauses = []
+        last = pks_left[-1]
+        rest = pks_left[:-1]
+        for i, pk in enumerate(rest):
```
Comment (nineinchnick) [inline datasette/utils/__init__.py]: if you want to make it more consistent with the original `compund_keys_after_sql`, use list comprehension here.

File: datasette/views/table.py [hunk 24]
```diff
@@ -543,29 +549,35 @@ async def data(
...
+            if name == "_sort" or name == "_sort_desc"
+        ]
+        sort_asc = []
+        sort_desc = []
+        if not sort:
+            if table_metadata.get("sort"):
+                sort.append(SortingOrder(table_metadata.get("sort"), "asc"))
+            if table_metadata.get("sort_desc"):
+                sort.append(SortingOrder(table_metadata.get("sort_desc"), "desc"))
+        for item in sort:
```
Comment (nineinchnick) [inline datasette/views/table.py]: nit: maybe you could create `sort_asc` and `sort_desc` using list comprehension.

File: datasette/views/table.py [hunk 25]
```diff
@@ -543,29 +549,35 @@ async def data(
...
-        if sort_desc:
-            if sort_desc not in sortable_columns:
-                raise DatasetteError(f"Cannot sort table by {sort_desc}")
-
-            order_by = f"{escape_sqlite(sort_desc)} desc"
+        #
+        # Getting sorting conditions from url and saving them in a list
+        #
+        sort = [
+            SortingOrder(escape_sqlite(value), "asc" if name == "_sort" else "desc")
```
Comment (nineinchnick) [inline datasette/views/table.py]: You call `escape_sqlite` later when building `order_by_list`. You might want to remove it here. Is there some test with special column names?

File: datasette/views/table.py [hunk 26]
```diff
@@ -581,18 +593,23 @@ async def data(
...
                 components = urlsafe_components(_next)
                 # If a sort order is applied, the first of these is the sort value
-                if sort or sort_desc:
-                    sort_value = components[0]
-                    # Special case for if non-urlencoded first token was $null
-                    if _next.split(",")[0] == "$null":
-                        sort_value = None
-                    components = components[1:]
+                if sort:
+                    sort_values = dict()
```
Comment (nineinchnick) [inline datasette/views/table.py:599]: If the order of elements matter here, you need to check what's the oldest Python version supported by Datasette - order in dicts is retained since 3.6

File: package-lock.json [hunk 27]
```diff
@@ -9,9 +9,9 @@
       }
     },
     "node_modules/prettier": {
-      "version": "2.2.1",
-      "resolved": "https://registry.npmjs.org/prettier/-/prettier-2.2.1.tgz",
-      "integrity": "sha512-PqyhM2yCjg/oKkFPtTGUojv7gnZAoG80ttl45O6x2Ug/rMJw4wcc9k6aaf2hibP7BGVCCM33gZoGjyvt9mm16Q==",
+      "version": "2.3.2",
```
Comment (nineinchnick) [inline package-lock.json]: If these changes are not related to the PR, you should revert them.

File: tests/test_api.py [hunk 28]
```diff
@@ -993,10 +1011,12 @@ def test_sortable_argument_errors(app_client):
     assert "Cannot sort table by badcolumn" == response.json["error"]
     response = app_client.get("/fixtures/sortable.json?_sort_desc=badcolumn2")
     assert "Cannot sort table by badcolumn2" == response.json["error"]
+    """ Test not needed anymore - Datasette should now support sorting by many conditions
```
Comment (nineinchnick) [inline tests/test_api.py]: Remove this then.

**Review by nineinchnick — APPROVED** [review]
